File: backend-selfhosted/app/core/obj_parser.py
"""
OBJ Parser

Parses Wavefront .obj text format -- the reliable fallback.
SketchUp exports .obj natively: File -> Export -> 3D Model -> OBJ.

This parser handles:
  - v  (vertex positions)
  - f  (faces, including n-gon faces with > 3 vertices)
  - g/o (group names -- used as labels)
  - Negative vertex indices
  - Faces with vertex/texture/normal index formats (v, v/vt, v/vt/vn, v//vn)

OBJ files have no standard unit. Coordinates are read as-is (no conversion).
The pipeline handles unit/scale detection separately.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

from .types import Face3D, Vec3, cross, normalize, sub


@dataclass
class ObjParseResult:
    faces: list[Face3D]
    warnings: list[str] = field(default_factory=list)


def parse_obj(text: str) -> ObjParseResult:
    warnings: list[str] = []
    vertices: list[Vec3] = []
    faces: list[Face3D] = []
    current_group: str | None = None

    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line or line[0] == "#":
            continue

        parts = line.split()
        keyword = parts[0]

        if keyword == "v":
            try:
                x = float(parts[1])
                y = float(parts[2])
                z = float(parts[3])
            except (IndexError, ValueError):
                continue
            if math.isfinite(x) and math.isfinite(y) and math.isfinite(z):
                vertices.append(Vec3(x, y, z))

        elif keyword in ("g", "o"):
            # Track group/object name for door/window detection.
            current_group = " ".join(parts[1:]) if len(parts) > 1 else None

        elif keyword == "f":
            idx_list: list[int] = []
            for i in range(1, len(parts)):
                token = parts[i].split("/")[0]
                try:
                    idx = int(token)
                except ValueError:
                    continue
                # OBJ indices are 1-based; negative = relative to end.
                if idx < 0:
                    idx = len(vertices) + idx + 1
                idx_list.append(idx - 1)  # convert to 0-based

            if len(idx_list) < 3:
                continue

            face_verts: list[Vec3] = []
            valid = True
            for idx in idx_list:
                if idx < 0 or idx >= len(vertices):
                    valid = False
                    break
                face_verts.append(vertices[idx])
            if not valid or len(face_verts) < 3:
                continue

            e1 = sub(face_verts[1], face_verts[0])
            e2 = sub(face_verts[2], face_verts[0])
            normal = normalize(cross(e1, e2))

            faces.append(Face3D(
                vertices=face_verts, normal=normal,
                group_name=current_group,
            ))

    if not faces:
        warnings.append("No faces found in the .obj file.")

    return ObjParseResult(faces=faces, warnings=warnings)

File: src/core/obj-parser.ts
// ============================================================================
// OBJ Parser
//
// Parses Wavefront .obj text format — the reliable fallback.
// SketchUp exports .obj natively: File → Export → 3D Model → OBJ.
//
// This parser handles:
//   - v (vertex positions)
//   - f (faces, including n-gon faces with > 3 vertices)
//   - g/o (group names — used as labels)
//   - Negative vertex indices
//   - Faces with vertex/texture/normal index formats (v, v/vt, v/vt/vn, v//vn)
// ============================================================================

import { type Face3D, type Vec3, cross, normalize, sub } from "./types";

export interface ObjParseResult {
  faces: Face3D[];
  warnings: string[];
}

export function parseObj(text: string): ObjParseResult {
  const warnings: string[] = [];
  const vertices: Vec3[] = [];
  const faces: Face3D[] = [];
  let currentGroup = "";

  const lines = text.split(/\r?\n/);

  for (const rawLine of lines) {
    const line = rawLine.trim();
    if (line.length === 0 || line[0] === "#") continue;

    const parts = line.split(/\s+/);
    const keyword = parts[0];

    if (keyword === "g" || keyword === "o") {
      // Group or object name — used as component ID for cutting sheets.
      currentGroup = parts.slice(1).join(" ") || "";
    } else if (keyword === "v") {
      // Vertex position.  OBJ coordinates are in the model's native unit.
      // SketchUp OBJ export uses inches by default; we convert to metres.
      // If the model was exported in metres, the user can adjust scale.
      const x = parseFloat(parts[1]);
      const y = parseFloat(parts[2]);
      const z = parseFloat(parts[3]);
      if (isFinite(x) && isFinite(y) && isFinite(z)) {
        vertices.push({ x, y, z });
      }
    } else if (keyword === "f") {
      // Face definition.
      const idxList: number[] = [];
      for (let i = 1; i < parts.length; i++) {
        // Format: v, v/vt, v/vt/vn, v//vn — we only need the first (vertex) index.
        const token = parts[i].split("/")[0];
        let idx = parseInt(token, 10);
        if (isNaN(idx)) continue;
        // OBJ indices are 1-based; negative = relative to end.
        if (idx < 0) idx = vertices.length + idx + 1;
        idxList.push(idx - 1); // convert to 0-based
      }

      if (idxList.length < 3) continue;

      // Gather vertices.
      const faceVerts: Vec3[] = [];
      let valid = true;
      for (const idx of idxList) {
        if (idx < 0 || idx >= vertices.length) { valid = false; break; }
        faceVerts.push(vertices[idx]);
      }
      if (!valid || faceVerts.length < 3) continue;

      // Compute normal from first 3 vertices.
      const e1 = sub(faceVerts[1], faceVerts[0]);
      const e2 = sub(faceVerts[2], faceVerts[0]);
      const normal = normalize(cross(e1, e2));

      faces.push({ vertices: faceVerts, normal, innerLoops: [], panelId: currentGroup || undefined });
    }
  }

  if (faces.length === 0) {
    warnings.push("No faces found in the .obj file.");
  }

  return { faces, warnings };
}
